src/CoreWCF.RabbitMQ/src/CoreWCF/Channels/RabbitmqTransportPump.cs

```diff
@@ -0,0 +1,78 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+using CoreWCF.Queue.Common;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using RabbitMQ.Client;
+using RabbitMQ.Client.Events;
+
+namespace CoreWCF.RabbitMQ.CoreWCF.Channels
+{
+    public class RabbitMqTransportPump : QueueTransportPump
+    {
+        private readonly ILogger<RabbitMqTransportPump> _logger;
+        private readonly IOptions<QueueOptions> _queueOptions;
+        private ConnectionFactory _factory;
+        private IConnection _connection;
+        private IModel _channel;
+
+        public RabbitMqTransportPump(ILogger<RabbitMqTransportPump> logger, IOptions<QueueOptions> options)
+        {
+            _logger = logger;
+            _queueOptions = options;
+        }
+
+        public override Task StartPumpAsync(QueueTransportContext queueTransportContext, CancellationToken token)
+        {
+            _factory = new ConnectionFactory { HostName = "localhost" };
+            _connection = _factory.CreateConnection();
+            _channel = _connection.CreateModel();
+
+            //Create a queue for messages destined to this service, bind it to the service URI routing key
+            string queue = _channel.QueueDeclare();
+            var exchange = "amq.direct";
+            // routing key begin with "/", for example: /hello
+            _channel.QueueBind(queue, exchange, _queueOptions.Value.QueueName, null);
+
+            var consumer = new EventingBasicConsumer(_channel);
+            consumer.Received += (_, ea) => { ConsumeMessage(ea, queueTransportContext); };
+            _channel.BasicConsume(queue, true, consumer);
+
+
+            return Task.CompletedTask;
+        }
+
+        public override Task StopPumpAsync(CancellationToken token)
+        {
+            _channel.Dispose();
+            _connection.Dispose();
+            return Task.CompletedTask;
+        }
+
+        private async void ConsumeMessage(BasicDeliverEventArgs eventArgs, QueueTransportContext queueTransportContext)
+        {
+            _logger.LogInformation("Receiving message from rabbitmq");
+            var reader = PipeReader.Create(new MemoryStream(eventArgs.Body.ToArray()));
+
```

Comment:
There's a lot of allocation and copying here. eventArgs.Body is a `ReadOnlyMemory<byte>` so calling ToArray allocates a new array and copies. Then MemoryStream is another allocation. There's a newer overload of PipeReader.Create which takes a `ReadOnlySequence<byte>` instead which avoids all this allocation. So can replace the code with tihs:
```c#
var reader = PipeReader.Create(new ReadOnlySequence<byte>(eventArgs.Body));
```
You might need to reference the 6.0 version of the nuget package System.IO.Pipelines, but that targets netstandard2.0 so should work before we require .net 6.0.